downloader/s3.go

```diff
@@ -0,0 +1,112 @@
+package downloader
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	sss "github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/simplifi/looking-glass/config"
+	"os"
+	"path"
+)
+
+type s3 struct {
+	awsSession session.Session
+	awsBucket  string
+	awsPrefix  string
+}
+
+func newS3(config config.DownloaderConfig) (Downloader, error) {
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	awsSess, err := createAwsSession(config.AwsKey, config.AwsSecret, config.AwsRegion)
+	if err != nil {
+		return nil, err
+	}
+	downloader := s3{
+		awsSession: *awsSess,
+		awsBucket:  config.AwsBucket,
+		awsPrefix:  config.AwsPrefix,
+	}
+
+	return &downloader, nil
+}
+
+func validateConfig(config config.DownloaderConfig) error {
+	if config.AwsKey == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
+	}
+	if config.AwsSecret == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsSecret")
+	}
+	if config.AwsRegion == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsSecret")
+	}
+	if config.AwsPrefix == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
+	}
+	if config.AwsBucket == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
+	}
+	return nil
+}
+
+func createAwsSession(awsKey string, awsSecret string, awsRegion string) (*session.Session, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(awsRegion),
+		Credentials: credentials.NewStaticCredentials(awsKey, awsSecret, ""),
+	})
+
+	return sess, err
+}
+
+func (s3s *s3) ListObjects() ([]string, error) {
+	var objects []string
+
+	resp, err := sss.New(&s3s.awsSession).
+		ListObjects(&sss.ListObjectsInput{
+			Bucket: aws.String(s3s.awsBucket),
+			Prefix: aws.String(s3s.awsPrefix),
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, obj := range resp.Contents {
+		objects = append(objects, *obj.Key)
+	}
+
+	return objects, nil
+}
+
+func (s3s *s3) GetObject(sourceObj string, targetPath string) error {
```

Comment:
Are we missing some linter checks? I thought that the linter required comments on public functions.

Reply:
Strange, there is no way to disable checks in the go linter, so I'm not sure why it isn't catching this.  I'll add some comments and will later look into the linter setup to see if I can figure out what is wrong.

I had thought about giving [revive](https://revive.run/) a try anyhow, maybe I'll do that as a follow on task.